File: python/day8.py
import sys
from collections import defaultdict


def print_antinodes(grid, antinode_locs):
    grid = grid.copy()
    for r, row in enumerate(grid):
        for c, elem in enumerate(row):
            if elem == "." and (r, c) in antinode_locs:
                grid[r][c] = "#"
    for line in grid:
        print("".join(line))
            

def part1():

    lines = parse()

    node_symbols = defaultdict(list)
    grid = lines
    ROWS, COLS = len(grid), len(grid[0])

    for r, row in enumerate(lines):
        for c, elem in enumerate(row):
            if elem != ".":
                node_symbols[elem].append((r, c))

    antinode_locations = list()
    for node_symbol, locs in node_symbols.items():
        for i in range(len(locs)):
            for j in range(i + 1, len(locs)):
                (x0, y0), (x1, y1) = sorted(((locs[i][0], locs[i][1]), (locs[j][0], locs[j][1])))
                a1_to_a0 = (x0 - x1, y0 - y1)
                a0_to_a1 = (x1 - x0, y1 - y0)
                antinode_locations.append((x0 + 2 * a0_to_a1[0], y0 + 2 * a0_to_a1[1]))
                antinode_locations.append((x1 + 2 * a1_to_a0[0], y1 + 2 * a1_to_a0[1]))

    within_bounds = lambda x: x[0] in range(ROWS) and x[1] in range(COLS)
    res = set(filter(within_bounds, antinode_locations))
    return len(res)
            

def part2():

    lines = parse()
    grid = lines

    node_symbols = defaultdict(list)
    ROWS, COLS = len(grid), len(grid[0])

    for r, row in enumerate(lines):
        for c, elem in enumerate(row):
            if elem != ".":
                node_symbols[elem].append((r, c))

    antinode_locations = set()
    for node_symbol, locs in node_symbols.items():
        for i in range(len(locs)):
            for j in range(i + 1, len(locs)):
                (x0, y0), (x1, y1) = sorted(((locs[i][0], locs[i][1]), (locs[j][0], locs[j][1])))
                a1_to_a0 = (x0 - x1, y0 - y1)
                a0_to_a1 = (x1 - x0, y1 - y0)

                get_loc1 = lambda steps: (x0 + (1 + steps) * a0_to_a1[0], y0 + (1 + steps) * a0_to_a1[1])
                steps = 1
                loc1 = get_loc1(steps)
                while loc1[0] in range(ROWS) and loc1[1] in range(COLS):
                    loc1 = get_loc1(steps)
                    antinode_locations.add(loc1)

                    steps += 1

                get_loc2 = lambda steps: (x1 + (1 + steps) * a1_to_a0[0], y1 + (1 + steps) * a1_to_a0[1])
                steps = 1
                loc2 = get_loc2(steps)
                while loc2[0] in range(ROWS) and loc2[1] in range(COLS):
                    loc2 = get_loc2(steps)
                    antinode_locations.add(loc2)

                    steps += 1

    within_bounds = lambda x: x[0] in range(ROWS) and x[1] in range(COLS)
    res = set(filter(within_bounds, antinode_locations)) | set(antenna_loc for node in node_symbols.values() for antenna_loc in node)
    return len(res)

def parse():
    with open(filename) as f:
        inp = f.read()
    lines = inp.split("\n")
    return list(map(list, lines))

if __name__ == "__main__":

    filename = sys.argv[1] if len(sys.argv) > 1 else "../data/day8.txt"

    print(f"{part1()=}")
    print(f"{part2()=}")

File: cpp/day8.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <map>
#include <set>
#include <algorithm>

// Read the grid from file
std::vector<std::vector<char>> read_grid(const std::string& filename) {
    std::ifstream infile(filename);
    std::vector<std::vector<char>> grid;
    std::string line;
    while (std::getline(infile, line)) {
        if (!line.empty()) grid.emplace_back(line.begin(), line.end());
    }
    return grid;
}

// Part 1: Find unique antinode locations
int part1(const std::vector<std::vector<char>>& grid) {
    int ROWS = grid.size(), COLS = grid[0].size();
    std::map<char, std::vector<std::pair<int, int>>> node_symbols;
    for (int r = 0; r < ROWS; ++r) {
        for (int c = 0; c < COLS; ++c) {
            if (grid[r][c] != '.') node_symbols[grid[r][c]].emplace_back(r, c);
        }
    }
    std::set<std::pair<int, int>> antinode_locations;
    for (const auto& [symbol, locs] : node_symbols) {
        for (size_t i = 0; i < locs.size(); ++i) {
            for (size_t j = i + 1; j < locs.size(); ++j) {
                auto [x0, y0] = locs[i];
                auto [x1, y1] = locs[j];
                if (std::tie(x0, y0) > std::tie(x1, y1)) std::swap(x0, x1), std::swap(y0, y1);
                int dx = x1 - x0, dy = y1 - y0;
                antinode_locations.emplace(x0 + 2 * dx, y0 + 2 * dy);
                antinode_locations.emplace(x1 - 2 * dx, y1 - 2 * dy);
            }
        }
    }
    // Only count those within bounds
    int count = 0;
    for (const auto& loc : antinode_locations) {
        int r = loc.first, c = loc.second;
        if (r >= 0 && r < ROWS && c >= 0 && c < COLS) ++count;
    }
    return count;
}

// Part 2: Find all antinode locations extended along the direction, plus all node locations
int part2(const std::vector<std::vector<char>>& grid) {
    int ROWS = grid.size(), COLS = grid[0].size();
    std::map<char, std::vector<std::pair<int, int>>> node_symbols;
    for (int r = 0; r < ROWS; ++r) {
        for (int c = 0; c < COLS; ++c) {
            if (grid[r][c] != '.') node_symbols[grid[r][c]].emplace_back(r, c);
        }
    }
    std::set<std::pair<int, int>> antinode_locations;
    for (const auto& [symbol, locs] : node_symbols) {
        for (size_t i = 0; i < locs.size(); ++i) {
            for (size_t j = i + 1; j < locs.size(); ++j) {
                auto [x0, y0] = locs[i];
                auto [x1, y1] = locs[j];
                if (std::tie(x0, y0) > std::tie(x1, y1)) std::swap(x0, x1), std::swap(y0, y1);
                int dx = x1 - x0, dy = y1 - y0;
                // Extend from (x0, y0) in direction (dx, dy)
                int steps = 1;
                while (true) {
                    int r = x0 + (1 + steps) * dx;
                    int c = y0 + (1 + steps) * dy;
                    if (r < 0 || r >= ROWS || c < 0 || c >= COLS) break;
                    antinode_locations.emplace(r, c);
                    ++steps;
                }
                steps = 1;
                while (true) {
                    int r = x1 - (1 + steps) * dx;
                    int c = y1 - (1 + steps) * dy;
                    if (r < 0 || r >= ROWS || c < 0 || c >= COLS) break;
                    antinode_locations.emplace(r, c);
                    ++steps;
                }
            }
        }
    }
    // Add all node locations
    for (const auto& [symbol, locs] : node_symbols) {
        for (const auto& loc : locs) antinode_locations.insert(loc);
    }
    // Only count those within bounds
    int count = 0;
    for (const auto& loc : antinode_locations) {
        int r = loc.first, c = loc.second;
        if (r >= 0 && r < ROWS && c >= 0 && c < COLS) ++count;
    }
    return count;
}

int main(int argc, char* argv[]) {
    std::string filename = argc > 1 ? argv[1] : "../data/day8.txt";
    auto grid = read_grid(filename);
    std::cout << "part1=" << part1(grid) << std::endl;
    std::cout << "part2=" << part2(grid) << std::endl;
    return 0;
} 